### PR TITLE
Build integrated OpenCL Linux wheels

### LINKED ISSUE #4684: Build Python wheels that support both GPU and CPU versions out of the box for non-Windows
Right now Python wheels from PyPI can be used to run training of LightGBM on both CPU and GPU (OpenCL-based implementation), thanks to #3144. It is done only for Windows, but can be implemented and for non-Windows systems as well, refer to 
> the static build technique also works on Linux, but will require a small upstream change in the build options for the Khronos ICD library. For now, I'm referencing my fork with that change in the Khronos build script. Depending on community feedback (that is, whether this build option would be useful on Linux too) I'll either pursue the upstream change in the Khronos library or pull out the Linux build code from this PR.
    https://github.com/microsoft/LightGBM/pull/3144#issue-630174938

https://github.com/microsoft/LightGBM/pull/3144#issuecomment-689182207

Also linking #2263 here as related.

To close this issue take #3144, #3660 as a reference and do the same for non-Windows systems.

I'm trying to push this forward due to the agreement on that OpenCL-based version is still important and shouldn't be abandoned after releasing new CUDA implementation: https://github.com/microsoft/LightGBM/pull/4630#discussion_r719122750.

### PR DESCRIPTION
**Pull request by jgiannuzzi**
This PR builds on @tpboudreau's excellent work in #3144 to add support for building integrated OpenCL wheels on Linux too. It also changes the CI to build this wheel instead of the regular CPU wheel, offering a better out-of-the-box experience for Linux users who want to use LightGBM with GPU support.

Fixes #4684

### CODE REVIEW

File: tests/python_package_test/conftest.py
Comment (jameslamb) [inline tests/python_package_test/conftest.py:3]: Could you tell us more about this? What specifically happened and how did you narrow it down to `dask` tests being the problem?

If library-loading order affects whether or not this works, then I worry that `lightgbm`'s users might experience whatever issue you're working around with this change to the tests.

Linking a possibly-related conversation: https://github.com/microsoft/LightGBM/issues/5106#issuecomment-1125634723

**Review by jameslamb — COMMENTED** [review]
Thank you very much for working on this! 

I personally am not very familiar with OpenCL and not qualified to review this. Hopefully @shiyu1994 @guolinke and @StrikerRUS will be able to provide you some feedback. I'm also pinging @huanzhang12 to possibly help.

**Comment by StrikerRUS** [comment]
@jgiannuzzi 
Thank you so much for working on this and for amazing PR!

> I personally am not very familiar with OpenCL and not qualified to review this.

Unfortunately, me too. But fortunately, I have an easy access to Ubuntu machine with NVIDIA GPU. So, I'll be able to independently test generated artifacts.

File: .vsts-ci.yml
Comment (StrikerRUS) [inline .vsts-ci.yml:49]: Could you please share the exact error message?

File: CMakeLists.txt
Comment (StrikerRUS) [inline CMakeLists.txt:523]: Could you please comment why this change was needed?

**Comment by StrikerRUS** [comment]
@tpboudreau @itamarst We'll really appreciate your inputs for this PR!

**Comment by StrikerRUS** [comment]
@jgiannuzzi Please update docs: https://github.com/microsoft/LightGBM/pull/3660/files#diff-e14b183376b1177323f6e7245d8ad64f2cd26f9638c129769d6b3c0ba4698dd5R26.

File: .vsts-ci.yml
Comment (jgiannuzzi) [reply .vsts-ci.yml:49]: This is the same error that we get when trying to do gpu-source with Python 3.8 on Ubuntu 14.04

File: tests/python_package_test/conftest.py
Comment (jgiannuzzi) [reply tests/python_package_test/conftest.py:3]: The issue is that the statically-linked OpenCL ICD loader is not able to load the AMD App SDK driver that we use for testing on the CPU. The call to `dlopen` fails with the error `cannot load any more object with static TLS`. This seems to only happen with the AMD App SDK. I have tested with an NVIDIA GPU and with the PoCL CPU implementation and it worked fine with both.

When running under Python 3.8 on Ubuntu 14.04, the AMD App SDK driver fails to load immediately. With Python 3.7, it only fails to load if the `dask` tests run first. I isolated that by running various combinations and orders of tests.

The whole issue seems to come from the AMD App SDK driver using statically allocated thread local storage (STATIC_TLS), which means that the amount of TLS objects is limited, and the library loading order can influence when the limit gets reached.

It's worth noting that both neither the NVIDIA OpenCL driver, nor the PoCL one use STATIC_TLS, so the wheel works fine with them under various versions of Python. I do not have access to the required hardware, but I have checked that the various ICDs available on Ubuntu 20.04 (Mesa, Beignet, and Intel) do not use STATIC_TLS either.

Comment (jgiannuzzi) [reply tests/python_package_test/conftest.py:3]: I have tried running the tests with the `gpu` and `bdist` tasks with PoCL instead of the AMD App SDK and they all succeed, even the `dask` tests that are currently disabled with `gpu`. I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK. It even enables testing of the Integrated OpenCL build on arm64!

I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

Comment (StrikerRUS) [reply tests/python_package_test/conftest.py:3]: @jgiannuzzi Thanks a lot for the detailed explanation!

> `cannot load any more object with static TLS`

Ah, I remember that annoying error! We've hit it already several times.
https://github.com/microsoft/LightGBM/pull/4893#issuecomment-1039204806
https://github.com/microsoft/LightGBM/pull/5022#issue-1146283008

> I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK.

Absolutely agree with you!

> It even enables testing of the Integrated OpenCL build on arm64!

Awesome!

> I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

We'll really appreciate PR with those changes. I believe we can have it prior to this PR because due to low maintenance activity of this project we prefer small focused of one feature/fix self-contained PRs. Otherwise, big PRs can have a long review cycle.

Comment (jgiannuzzi) [reply tests/python_package_test/conftest.py:3]: Great, I opened #5282!

Comment (StrikerRUS) [reply tests/python_package_test/conftest.py:3]: We install AMD OpenCL for Linux here
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/setup.sh#L68-L82
and for Windows here
https://github.com/microsoft/LightGBM/blob/master/.ci/install_opencl.ps1
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/test_windows.ps1#L77-L78

Comment (StrikerRUS) [reply tests/python_package_test/conftest.py:3]: > Great, I opened https://github.com/microsoft/LightGBM/pull/5282!

Thanks a lot!

File: .ci/test.sh
Comment (StrikerRUS) [inline .ci/test.sh:165]: >  It even enables testing of the Integrated OpenCL build on arm64!

Can we try to integrate OpenCL into `aarch64` Linux wheels?
https://github.com/microsoft/LightGBM/blob/ab9236baa06c3f071acf6b45a771c2a09661f146/.vsts-ci.yml#L158
Or maybe you've already tried to do it?

File: .ci/test.sh
Comment (StrikerRUS) [inline .ci/test.sh:167]: Please update docs: https://github.com/microsoft/LightGBM/pull/3660/files#diff-e14b183376b1177323f6e7245d8ad64f2cd26f9638c129769d6b3c0ba4698dd5R26.

File: .ci/test.sh
Comment (jgiannuzzi) [reply .ci/test.sh:165]: I have started to work on it. I had tested successfully with ubuntu 20.04, but then I noticed that we use CentOS in the CI. I hope to add a commit enabling aarch64 tomorrow.

File: .ci/test.sh
Comment (jgiannuzzi) [reply .ci/test.sh:167]: Will do, thanks for the reminder!

File: .ci/test.sh
Comment (StrikerRUS) [reply .ci/test.sh:165]: Sounds great!

Comment (jgiannuzzi) [reply .ci/test.sh:163]: Enabled in 6a08e4f and c4e0533

File: .ci/test.sh
Comment (jgiannuzzi) [reply .ci/test.sh:165]: Done in 594a711

**Comment by jgiannuzzi** [comment]
@StrikerRUS I have addressed your remaining comments.

@tpboudreau Could you please take a look? For the context, #5282 got merged before this one, switching the OpenCL CPU implementation to [PoCL](http://portablecl.org) on Linux.

File: .ci/test.sh
Comment (StrikerRUS) [reply .ci/test.sh:163]: Unfortunately, `test_dual.py` fails with `lightgbm.basic.LightGBMError: No OpenCL device found` on `aarch64`.
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12929&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9&l=977

File: .ci/test.sh
Comment (StrikerRUS) [reply .ci/test.sh:165]: Thanks a lot! But I think that `integrated-opencl` build option shouldn't be exposed to the public interface.
Could you please revert the following added sections?

image

File: .ci/test.sh
Comment (StrikerRUS) [reply .ci/test.sh:163]: Maybe [`gpu_platform_id`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#gpu_platform_id) and [`gpu_device_id`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#gpu_device_id) params can help?..

File: tests/python_package_test/test_dual.py
Comment (StrikerRUS) [inline tests/python_package_test/test_dual.py:30]: Could you please explain why we don't set `gpu_use_dp=True` on aarch64?

**Comment by StrikerRUS** [comment]
We've recently moved PoCL installation from the CI runtime (`.ci/setup.sh`) to the Docker creation phase with the aim to reduce overall CI time: https://github.com/guolinke/lightgbm-ci-docker/pull/26 and #5286. That process introduced some merge conflicts in this PR. Sorry about that! Could you please resolve those conflicts?

**Comment by StrikerRUS** [comment]
Good news!

I just checked generated as CI artifact wheel file on my Ubuntu machine with NVIDIA GPU. I simply installed it with `pip install *.whl` and after that `device='gpu'` parameter allowed me to utilize my GPU.

```python
>>> Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53)
>>> [GCC 9.4.0] on linux

import numpy as np
import lightgbm as lgb

X = np.random.random((10_000, 200))
y = np.random.random(10_000)

est = lgb.LGBMRegressor(n_estimators=5000).fit(X, y)
>>> [LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.009667 seconds.
>>> You can set `force_col_wise=true` to remove the overhead.
>>> [LightGBM] [Info] Total Bins 51000
>>> [LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 200
>>> [LightGBM] [Info] Start training from score 0.502931

est = lgb.LGBMRegressor(n_estimators=5000, device='gpu').fit(X, y)
>>> [LightGBM] [Info] This is the GPU trainer!!
>>> [LightGBM] [Info] Total Bins 51000
>>> [LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 200
>>> [LightGBM] [Info] Using GPU Device: Tesla V100-SXM2-32GB, Vendor: NVIDIA Corporation
>>> [LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
>>> [LightGBM] [Info] GPU programs have been built
>>> [LightGBM] [Info] Size of histogram bin entry: 8
>>> [LightGBM] [Info] 200 dense feature groups (1.91 MB) transferred to GPU in 0.004207 secs. 0 sparse feature groups
>>> [LightGBM] [Info] Start training from score 0.502931
```

image

**Comment by StrikerRUS** [comment]
@jgiannuzzi Hey! Maybe we can merge this PR without the support for aarch64 and add it later in a separate PR?

**Comment by jgiannuzzi** [comment]
Hey @StrikerRUS, I'm very sorry for not having updated this PR yet. I have had a fix for aarch64 for a while but never had the time to update the PR. I will try to do it this week so we can finally have those integrate OpenCL Linux wheels!

**Comment by StrikerRUS** [comment]
@jgiannuzzi No problem. Thanks a lot for all your hard work!

**Comment by ethanmdev** [comment]
So keen for this PR to get merged! @jgiannuzzi

File: tests/python_package_test/test_dual.py
Comment (jgiannuzzi) [reply tests/python_package_test/test_dual.py:30]: The reason is that double-precision floats are not supported with PoCL on aarch64.

File: .ci/test.sh
Comment (jgiannuzzi) [reply .ci/test.sh:170]: Done!

File: .ci/test.sh
Comment (jgiannuzzi) [reply .ci/test.sh:168]: This has been fixed in 0e25310. The issue was a wrong condition for detecting aarch64 (I was using the environment variable `ARCH` that was never set).

**Comment by jgiannuzzi** [comment]
Hey @StrikerRUS, I finally got around to updating this PR! I believe that I took into account all of your comments and I also fixed the support for aarch64.

It's taking 30 additional minutes because of building PoCL under emulation, which I believe we could solve in a separate PR by pre-building it (like we do for Ubuntu 14.04). I'm happy to take a shot at it later.

File: tests/python_package_test/test_dual.py
Comment (StrikerRUS) [reply tests/python_package_test/test_dual.py:30]: Ah, I see! Thanks for the explanation.

Could you please add comment "double-precision floats are not supported with PoCL on aarch64" in the code to make it clear for everyone reading the code?

**Comment by StrikerRUS** [comment]
> It's taking 30 additional minutes because of building PoCL under emulation, which I believe we could solve in a separate PR by pre-building it (like we do for Ubuntu 14.04). I'm happy to take a shot at it later.

That would be a little bit trickier because right now we use docker image from pypa org:
https://github.com/microsoft/LightGBM/blob/e063dad261d06049a461c3e2a364350e2b20b9f7/.vsts-ci.yml#L192
I think we can switch to using an our own one and referencing the current one as a base `FROM`.

File: tests/python_package_test/test_dual.py
Comment (jgiannuzzi) [reply tests/python_package_test/test_dual.py:31]: Done!

**Review by StrikerRUS — APPROVED** [review]
Thank you so much for this PR!
All changes look good to me!

**Comment by StrikerRUS** [comment]
> I think we can switch to using an our own one and referencing the current one as a base `FROM`.

Already started preparing a PR for this.
https://github.com/guolinke/lightgbm-ci-docker/compare/dev
Will continue after merging this PR.

**Comment by jgiannuzzi** [comment]
Hi @jameslamb and @StrikerRUS, are you happy to merge this?

I'm looking forward to having nightly builds of LightGBM with GPU support on Linux, so let me know if there's anything else I can do!

**Comment by StrikerRUS** [comment]
@jameslamb @jmoralez @guolinke @shiyu1994 Could you please help with a second review for this PR?

**Comment by jameslamb** [comment]
I can review by the end of this week. Sorry for the delay @jgiannuzzi , and thanks for the awesome work!

**Comment by shiyu1994** [comment]
Thanks for working on this. This is super helpful. I'll come back to review this in this week, if by then no further review is provided.

**Comment by jameslamb** [comment]
I'll be able to review today, sorry I didn't get to it yet!

File: .ci/setup.sh
Comment (jameslamb) [inline .ci/setup.sh:86]: Why do these `yum install` blocks need to be split up? If there isn't a specific reason, could you please combine them into a single command (with alphabetized list of libraries)?

File: .ci/setup.sh
Comment (jameslamb) [inline .ci/setup.sh:90]: In the time since this PR was first opened, PoCL has put up a new major release (v3.0, July 2022): https://github.com/pocl/pocl/releases.

Could you try upgrading to that, to take advantage of bugfixes and improvements in that project? I think that right now, while this support is new and we're actively testing it, is a great time to try to use the newest version possible.

Or, if that's a large amount of effort, after merging this would you please write up a new issue documenting the work to upgrade to that new version of PoCL?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much for this! I left two suggestions for your consideration.

@StrikerRUS since we haven't been successful getting any other reviews with more OpenCL understanding to review...could you test the generated wheels one more time? There have been some additional commits since you last tested (https://github.com/microsoft/LightGBM/pull/5252#issuecomment-1159807098).

I'm also interested in understanding how much larger the package's wheels will be as a result of this change. Large wheels can introduce some new classes of problems for users (for example, those using function-as-a-service services that have fairly strict limits on disk usage, see https://github.com/pandas-dev/pandas/issues/30741#issuecomment-752751976). I'll wait until another commit is pushed to check that (since the previous Azure DevOps build is gone now).

File: .ci/setup.sh
Comment (jgiannuzzi) [reply .ci/setup.sh:86]: This is because installing `epel-release` adds a yum repository required for installing the packages in the second command. This can unfortunately not be done in a single command.

File: .ci/setup.sh
Comment (jgiannuzzi) [reply .ci/setup.sh:90]: Thanks for letting me know, I'll take a look at upgrading to PoCL v3.0!

For this specific PR, I have only aligned linux aarch64 to also use PoCL like we do on linux x86-64. I would thus suggest leaving the upgrade out of the scope of this PR.

**Comment by jgiannuzzi** [comment]
> I'm also interested in understanding how much larger the package's wheels will be as a result of this change. Large wheels can introduce some new classes of problems for users (for example, those using function-as-a-service services that have fairly strict limits on disk usage, see [pandas-dev/pandas#30741 (comment)](https://github.com/pandas-dev/pandas/issues/30741#issuecomment-752751976)). I'll wait until another commit is pushed to check that (since the previous Azure DevOps build is gone now).

@jameslamb I have rebased my branch so that the CI can run again

**Comment by jameslamb** [comment]
Thanks for that! I'll take a look shortly.

In the future, when contributing on this project please use merge commits instead of rebasing + force-pushing. Rebasing and force-pushing makes it more difficult for reviewers to see what was changed between review cycles, and breaks the link to our CI systems (which makes it much more difficult to trace back to "what happened in CI as of when this commit was pushed").

This project squashes all commits into 1 when merging, so it isn't necessary to keep a linear commit on PR branches.

File: .ci/setup.sh
Comment (jameslamb) [reply .ci/setup.sh:90]: Ok nice, makes sense! Once we merge this PR, would you please document a feature request like "upgrade to PoCL v3.0" in this project's issue tracker? Just so we don't forget about. it.

File: .ci/setup.sh
Comment (jameslamb) [reply .ci/setup.sh:86]: ahhhh ok that makes sense. Thanks for the explanation.

**Comment by jgiannuzzi** [comment]
Sorry about the rebase! I have reverted to the previous commit and merged `master` instead.

@jameslamb could you please take a look again once the CI has run?

**Review by jameslamb — COMMENTED** [review]
Thanks very much for your patience and excellent contribution @jgiannuzzi !

We've finally fixed the CI issues in this project (#5510), and I'd like to help move this PR forward.

I don't have any other changes to recommend, but I would like to see CI run successfully and check the artifact size before approving. Can you please merge the latest changes from `master` into this branch?

I'm still hoping @StrikerRUS will be able to test again (https://github.com/microsoft/LightGBM/pull/5252#pullrequestreview-1113112842), but it's been a few weeks since I asked, so I think he may be unavailable for a while. If we don't hear from him in the next few days, I can spin up a box in AWS and test with something like the approach I took in https://github.com/microsoft/LightGBM/pull/5369#pullrequestreview-1064356007.

@guolinke @shiyu1994 could one of you help with a review as well?

**Comment by jgiannuzzi** [comment]
> Can you please merge the latest changes from `master` into this branch?

@jameslamb done!

**Comment by jgiannuzzi** [comment]
@jameslamb if it helps in any way, I have access to NVIDIA hardware (both for x86_64 and aarch64) and I'd be happy to help execute any tests you'd like to do. I can even do it with you through screen sharing if you'd like.

**Comment by jgiannuzzi** [comment]
> I'm sorry, but I'm still struggling to get the artifacts from Azure DevOps. Either the retention for those artifacts is _very_ brief or I'm just doing something wrong.

I just checked and artifact publishing is not enabled for pull requests! I'm afraid you'll have to build the wheel locally to test it.

> Could you just tell me how large the wheels built this way are in your local setup? That'll give me a good enough idea to know if wheel size is something we need to optimize more for here.

I haven't done this in a while — I'll try to do this ASAP and report the size here.

**Comment by jgiannuzzi** [comment]
@jameslamb the x86_64 package size with built-in OpenCL is 2.2MB.

Here is how I simulated a CI run to produce the wheel:

```bash
docker run --rm -ti \
--entrypoint bash \
-v $PWD/artifacts:/artifacts \
-e AZURE=true \
-e BUILD_DIRECTORY=/vsts/lightgbm \
-e BUILD_SOURCESDIRECTORY=/vsts/lightgbm \
-e BUILD_ARTIFACTSTAGINGDIRECTORY=/artifacts \
-e COMPILER=gcc \
-e CONDA_ENV=test-env \
-e LGB_VER=3.3.2.99 \
-e OS_NAME=linux \
-e PATH=/opt/miniforge/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
-e PRODUCES_ARTIFACTS=true \
-e PYTHON_VERSION=3.8 \
-e SETUP_CONDA=false \
-e TASK=bdist \
-e OMP_NUM_THREADS=4 \
-e POCL_MAX_PTHREAD_COUNT=4 \
lightgbm/vsts-agent:ubuntu-14.04 \
-c "git clone https://github.com/jgiannuzzi/lightgbm.git --recursive --depth 1 --branch linux-gpu-wheel && cd lightgbm && .ci/setup.sh && .ci/test.sh && chown -R $(id -u):$(id -g) /artifacts"
```
The last 2 environment variables are only needed on a machine with many cpu cores (mine has 256) to avoid some tests taking a huge amount of time. You can skip them if you have less than 16 cores (the CI does _not_ have them).

The wheel can be found under `artifacts` in the current working directory.

**Comment by jgiannuzzi** [comment]
Hi @jameslamb @guolinke @shiyu1994! Is there anything I can do to help with getting this change merged?

**Comment by jameslamb** [comment]
> the x86_64 package size with built-in OpenCL is 2.2MB

Amazing 🙌 

I was worried you were going to say like "125MB". 2.2MB compressed is great, that's the same order of magnitude as LightGBM's existing wheels (https://pypi.org/project/lightgbm/#files).

I'm sorry it's taken so long for me to test this out. We have been dealing with a serious lack of maintainer availability + a string of difficult CI issues (#5549, #5562, #5569).

Those CI issues are now resolved, and testing this is my top priority.

In the interim, can you please merge the latest changes from `master` into this branch?

Thank you so much for being patient with us. We really really appreciate this contribution and hope you'll continue contributing to LightGBM in the future.

**Comment by jameslamb** [comment]
closing and re-opening to trigger CI

**Comment by jameslamb** [comment]
I tried building this in CI last night (just pushing this branch from a fork to a LightGBM branch so it'd produce artifacts I could download and test with), and unfortunately I saw one failure on the `QEMU_multiarch bdist` build.

The `test_cpu_and_gpu_work()` test failed with the following.

> lightgbm.basic.LightGBMError: No OpenCL device found

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13872&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9))

I just pushed https://github.com/microsoft/LightGBM/pull/5252/commits/a9f02d72db166eff8434deed5e08e729db885d09 updating this to latest `master`. Let's see if that happens here.

I'm so sorry @jgiannuzzi , I'm trying to test this but LightGBM's CI is not in a good state right now. I'm trying to fix it as fast as I can.

File: .ci/setup.sh
Comment (jameslamb) [reply .ci/setup.sh:90]: I created #5596 so we don't forget about this

**Comment by jameslamb** [comment]
I realized that in the new `manylinux_2_28_x86_64` image used to build `x86_64` linux wheels, a too-old OpenCL was being found, and as a result `pocl` didn't register itself in `/etc/OpenCL/vendors`.

Fixed that in https://github.com/guolinke/lightgbm-ci-docker/pull/29, which looks like it fixed https://github.com/microsoft/LightGBM/pull/5252#issuecomment-1321192047.

([ci build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13934&view=results))

I'll try fixing the other two CI failures tonight, they're almost certainly my fault. Sorry about that @jgiannuzzi , I'm doing my best to get this merged.

**Comment by guolinke** [comment]
I am not sure why the  "guolinke/lightgbm-ci-docker#29"  will auto close this

**Comment by jameslamb** [comment]
> I am not sure why the "https://github.com/guolinke/lightgbm-ci-docker/pull/29" will auto close this

oh strange! Probably in the way I phrased "fix" in that PR's description. Sorry about that and thanks for re-opening it.

**Review by jameslamb — APPROVED** [review]
Alright I've delayed this long enough, going to approve based on the passing CI and all I've learned about PoCL and this PR's other additions while working through #5580.

I'd still like to try these wheels again on different GPUs from a cloud provider, but that can be done later.

@guolinke can you look one more time? If you approve, please merge this (I'll be traveling for the next few days).

@jgiannuzzi thank you SO MUCH for this awesome contribution, and the other excellent contributions you've made to LightGBM along the way. I've learned a lot from you and it's been great working with you. I hope you'll consider contributing more to LightGBM in the future 😁

**Review by guolinke — APPROVED** [review]
Thank you!

**Comment by jgiannuzzi** [comment]
Thank you @jameslamb and @guolinke, I'm looking forward to having daily builds with GPU support!

I'm sorry I wasn't available to help earlier with the many CI woes, and I'm certainly looking forward to contributing again to LightGBM in the future!

**Pull request merged**
